### PR TITLE
Makes torch.floor_divide consistent with Python and NumPy and torch.Tensor.__rtruediv__ consistent with div's new true division behavior

### LINKED ISSUE #43874: floor_divide truncates instead of flooring
## 🐛 Bug

PyTorch's floor_divide (//) operator doesn't match Python's behavior on negative numbers. Python floors (rounds down) while PyTorch truncates (rounds towards zero).

## To Reproduce

```Python
import torch
import numpy as np

print("python -3//2 =", -3 // 2)
print("numpy -3//2 =", np.array((-3)) // 2)
print("pytorch -3//2 =", torch.tensor([-3]) // 2)

# output:
# python -3//2 = -2
# numpy -3//2 = -2
# pytorch -3//2 = tensor([-1])
```

## Expected behavior

Expect PyTorch's floor_divide to match Python // operator by flooring division results instead of truncating them.

## Environment
```

 - PyTorch Version (e.g., 1.0): master
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): 
 - Python version: Python 3.7.5
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information:
```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser

### PR DESCRIPTION
**Pull request by mruberry**
**BC-breaking note**

torch.floor_divide now correctly takes the floor of its quotient instead of truncating it. In other words, torch.floor_divide now correctly rounds the quotient down instead of towards zero. This affects floor division where the dividend and divisor have different signs.

In PyTorch 1.6:

```
torch.floor_divide(-5, 3) 
: tensor(-1)
```

But In PyTorch 1.7:

```
torch.floor_divide(-5, 3) 
: tensor(-2)
```

torch.Tensor.\_\_rtruediv\_\_ has been updated to perform true division like `torch.true_divide`, instead of mimicking C++-style integer division when given an integer tensor. 

In PyTorch 1.6:

```
5 / torch.tensor(3)
: tensor(1)
```

In PyTorch 1.7:

```
5 / torch.tensor(3)
: tensor(1.6667)
```

**PR Summary**

Fixes https://github.com/pytorch/pytorch/issues/43874. 

torch.floor_divide was performing...

1) a true division followed by a truncation for float types
2) integer division in C++ for integer types

Unfortunately, this is inconsistent with both Python and NumPy, where floor division is a true division followed by a floor (as its name would suggest). For example in Python:

```
-3 // 2 = -2
```

While in C++:

```
-3 / 2 = -1
```

Note that floor division between a positive divisor and dividend is unaffected by this change, since truncating positive values is equivalent to flooring them. 

An unfortunate side affect of this change is that it leaves C++ style integer division inaccessible to users. A workaround to mimic the C++ behavior is `torch.trunc(a / b)`.

Additionally, the behavior of reverse true division on tensors has been made consistent with torch.true_divide. This change was included in this PR because it's very simple and it greatly facilities writing comprehensive division tests.

### CODE REVIEW

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 1]
```diff
@@ -200,7 +200,7 @@ void index_put_accum_kernel(Tensor & self, TensorList indices, const Tensor & va
       using device_ptr = thrust::device_ptr<int64_t>;
       const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Indexing.cu:202]: See if there's a comment we can make here

Comment (mruberry) [reply aten/src/ATen/native/cuda/Indexing.cu:202]: Test failure - did not investigate

File: torch/overrides.py [hunk 2]
```diff
@@ -178,6 +178,9 @@ def get_ignored_functions() -> Set[Callable]:
         Tensor._make_subclass,
         Tensor.stride,
         Tensor.unflatten,
+        torch._c_style_div,
```
Comment (gchanan) [inline torch/overrides.py:181]: why are these here?  Seems fine, but there are tensor-likes.

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -2728,20 +2730,26 @@ def merge_dicts(*dicts):
...
 
 .. math::
     \text{{out}}_i = \left\lfloor \frac{{\text{{input}}_i}}{{\text{{other}}_i}} \right\rfloor
 
 """ + r"""
+
+Supports :ref:`broadcasting to a common shape <broadcasting-semantics>`,
+:ref:`type promotion <type-promotion-doc>`, and integer and float inputs.
+
+.. note:: Floor division in PyTorch is like floor division in Python
```
Comment (gchanan) [inline torch/_torch_docs.py:2746]: nit: NumPy.

Comment (gchanan) [inline torch/_torch_docs.py:2746]: nit: torch.floor_divide performs floor division as in Python and NumPy.  This differs from integer division in C++, which truncates instead of flooring the quotient.  E.g. in Python, -5 // 2 == -3, but in C++ -5 // 2 = -2.

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -2381,7 +2381,9 @@ def merge_dicts(*dicts):
 .. warning::
     Integer division using div is no longer supported, and in a future release
     div will perform true division as in Python 3. Use :func:`torch.true_divide`
-    or :func:`torch.floor_divide` (// in Python), instead.
+    or :func:`torch.floor_divide` (// in Python), instead. To mimic the historic
```
Comment (gchanan) [inline torch/_torch_docs.py:2384]: should it be `//` instead of //?

Comment (gchanan) [reply torch/_torch_docs.py:2384]: nit: to mimic historical integer division (C-style), use...

File: torch/_torch_docs.py (hunk 3 above)
Comment (gchanan) [reply torch/_torch_docs.py:2746]: perhaps its better to refer to this as C-style, assuming C and C++ behave the same.

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -7704,8 +7704,25 @@
   python_module: linalg
   variants: function
 
+## Undocumented functions intended for internal use only
+- func: _c_style_div(Tensor self, Tensor other) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml:7708]: we technically don't need the Tensor overloads, right?

Comment (gchanan) [reply aten/src/ATen/native/native_functions.yaml:7708]: nor do we need these to be functions.

File: aten/src/ATen/native/BinaryOps.cpp [hunk 6]
```diff
@@ -213,29 +215,22 @@ Tensor& true_divide_(Tensor& self, const Tensor& divisor) {
...
+//   type promotion.
 Tensor& floor_divide_out(Tensor& result, const Tensor& self, const Tensor& other) {
-  auto iter = TensorIterator::binary_op(result, self, other);
-  div_stub(iter.device_type(), iter);
-
-  if (result.is_floating_point()) {
-    result.trunc_();
-  }
-
+  const auto r = self.true_divide(other).floor_();
```
Comment (gchanan) [inline aten/src/ATen/native/BinaryOps.cpp:225]: does `at::true_divide_out(result, self, other).floor_()` not work?  So you don't need to copy.

File: aten/src/ATen/native/BinaryOps.cpp [hunk 7]
```diff
@@ -213,29 +215,22 @@ Tensor& true_divide_(Tensor& self, const Tensor& divisor) {
...
-
-  div_stub(iter.device_type(), iter);
-
-  auto out = iter.output();
-  if (out.is_floating_point()) {
-    out.trunc_();
-  }
-
-  return out;
+  const auto result_dtype = at::native::result_type(TensorList{self, other});
```
Comment (gchanan) [inline aten/src/ATen/native/BinaryOps.cpp:232]: why don't you call floor_divide_out?

Also, isn't true_divide guaranteed to give the same result type?

File: aten/src/ATen/native/BinaryOps.cpp [hunk 8]
```diff
@@ -1051,5 +1046,30 @@ TORCH_LIBRARY_IMPL(aten, Meta, m) {
...
+Tensor _c_style_div(const Tensor& self, const Tensor& other) {
+  Tensor result;
+  auto iter = TensorIterator::binary_op(result, self, other);
+  div_stub(iter.device_type(), iter);
+  return iter.output();
+}
+
+Tensor& _c_style_div_(Tensor& self, const Tensor& other) {
+  auto iter = TensorIterator::binary_op(self, self, other);
+  div_stub(iter.device_type(), iter);
```
Comment (gchanan) [inline aten/src/ATen/native/BinaryOps.cpp:1062]: nit: calling this div_stub doesn't seem great.

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (mruberry) [reply aten/src/ATen/native/native_functions.yaml:7708]: We can prune this hidden function back to just the needed version, sure.

File: aten/src/ATen/native/BinaryOps.cpp (hunk 6 above)
Comment (mruberry) [reply aten/src/ATen/native/BinaryOps.cpp:225]: The type of result may not be copyable to by the result type of true divide out. This could be tested for, however.

File: aten/src/ATen/native/BinaryOps.cpp (hunk 7 above)
Comment (mruberry) [reply aten/src/ATen/native/BinaryOps.cpp:232]: We could call out.

true_divide will always return a float or complex type because it always promotes to float, but floor divide can return integer types because it participates in regular type promotion.

File: aten/src/ATen/native/BinaryOps.cpp (hunk 8 above)
Comment (mruberry) [reply aten/src/ATen/native/BinaryOps.cpp:1062]: We could follow-up in a separate pr and change the name.

File: torch/overrides.py (hunk 2 above)
Comment (mruberry) [reply torch/overrides.py:181]: I just didn't want to let people override the secret function: they're not supposed to be using it.

File: torch/_torch_docs.py (hunk 3 above)
Comment (mruberry) [reply torch/_torch_docs.py:2746]: SG.

File: torch/_torch_docs.py (hunk 4 above)
Comment (mruberry) [reply torch/_torch_docs.py:2384]: SG.

**Comment by gchanan** [comment]
can you write specific examples showing both changes for the release note?  In "Version 1.6, Version 1.7" style like the release notes, preferrably.

File: aten/src/ATen/native/BinaryOps.cpp [hunk 9]
```diff
@@ -194,50 +197,76 @@ Tensor& true_divide_(Tensor& self, Scalar divisor) {
...
+  // Computes the output dtype
+  auto computation_dtype = at::native::result_type(TensorList{self, other});
+
+  TORCH_CHECK(computation_dtype != ScalarType::Bool,
+    "torch.floor_divide is not implemented for scalar type bool!");
+
+  // NOTE: floordiv_stub handles the special case of integer floor division
+  if (c10::isIntegralType(computation_dtype, /*include_bool=*/false)) {
+    auto iter = TensorIterator::binary_op(result, self, other);
+    floordiv_stub(iter.device_type(), iter);
```
Comment (gchanan) [inline aten/src/ATen/native/BinaryOps.cpp:218]: nit: can you rename this floordiv_integral_stub or something?  Looking at floordiv_stub, I would have expected it implements floordiv.

Comment (mruberry) [reply aten/src/ATen/native/BinaryOps.cpp:218]: Sure.

**Review by gchanan — APPROVED** [review]

**Comment by BowenBao** [comment]
hi @mruberry - is this merged? i could not find this commit in master. I have the PR to update ONNX export to stay consistent #45161.

**Comment by mruberry** [comment]
> hi @mruberry - is this merged? i could not find this commit in master. I have the PR to update ONNX export to stay consistent #45161.

Nope; we decided to go through a formal deprecation process for the floor_divide behavior. I'll update the issue. Thanks for reminding me.

**Comment by BowenBao** [comment]
@mruberry - ok, i'm closing the other PR for now.

**Comment by mruberry** [comment]
> @mruberry - ok, i'm closing the other PR for now.

Thank you, @BowenBao. Sorry for the disruption. We found some users who rely on floor division's current behavior and don't want to break them.
